### DISCUSSION TITLE
Converting Python lists to JS arrays

### DISCUSSION THREAD

**Discussion by cfry** [pyodide/pyodide#4531] [Q&A]
In my program using pyodide, I can eval the following Python code:
js.Dex.m_joint(5, 10, 25) 
and the correct args are passed to my Javascript function Dex.m_joints (integers)
This function can also take a JS array as its first argument.
When I call, from python,
js.Dex.m_joint([5, 10, 25]) 
instead of a js array getting passed to my function, a pyProxy is passed instead.
Please tell me how I can get an array passed in.

**Comment by hoodmane** [comment]
```py
from pyodide.ffi import to_js

js.some_func(to_js([1, 2, 3]))
```

**Comment by cfry** [comment]
OK this works.
It seems so useful, why wouldn't this best
be done automatically behind the scenes
so a python program could just write
js.Dex.m_xyz([0.1, 0.2, 0.3])

instead of
from pyodide.ffi import to_js
js.Dex.m_xyz(to_js([0.1, 0.2, 0.3]))
?
I suspect this is effectively going on for some
other data structures.
Should I make a "feature request" ?

**Reply by hoodmane** [reply]
This behavior is intentional. Consider the case where the code mutates the list:
```js
function f(l) {
   l.push(5);
}
pyodide.globals.set("f", f)
pyodide.runPython(`
  l = [1, 2, 3]
  f(l)
  print(l) # prints [1, 2, 3, 5]
`);
```
Your proposal makes this impossible. Once upon a time it did work as you suggest, see #900 for discussion about why we changed it.

**Comment by cfry** [comment]
I am more than impressed with your helpful,
detailed and fast (even on a weekend!) responses to my questions.
I also appreciated that you're taking care of not just a round trip
but 1.5 round trips for data passed between python and JS.

Are you or someone intimately familiar with Piodide
available for consulting?
____
From the #900 discussion,
Admittedly I didn't understand all of it,
but sounded like it might still be reasonable to convert
tuples (10, 20, 30) , but not lists,    into JS arrays.
Is that still a possibility?
_____
With the proxy sent back to JS for that
(hopefully) array arg,
I tried calling the_proxy.toJs()
on it but got an error:
Uncaught Error: This borrowed proxy was automatically destroyed at the end
of a function call. Try using create_proxy or create_once_callable.

BUT: I didn't create the proxy (at least explicitly in my code)
nor did I explicitly destroy it.
Perhaps there's some other way I can extract a
js array from that returned proxy.

Otherwise, it seems like I must convert that list/tuple
before that arg leaves my python environment, right?
